### ISSUE TITLE
Script.resource_name always contains blank string

### ISSUE THREAD

**Issue by damarindra** [godotengine/godot#13358]
**Operating system or device, Godot version, GPU Model and driver (if graphics related):**
a26b36bec2e99a128b7a61ff9fa1e2e46148f547

**Issue description:**
Whenever you try to .get_script().resource_name, the variable only contains blank string. However resource_path always return valid path.

**Steps to reproduce:**
1. Attach script with this line code
`print(get_script().resource_name)`
2. See on the console, doesn't print anything

**Comment by damarindra** [comment]
Run this demo, resouce_name will not print anything, but resource_path is print the right path
Demo : [resource_name_blank_string.zip](https://github.com/godotengine/godot/files/1513260/resource_name_blank_string.zip)

**Comment by eska014** [comment]
Is resource name used at all internally? Otherwise the field should just be removed

**Comment by vnen** [comment]
If you look at the script on the Inspector, you'll see that the resource name is empty. Since GDScript files contain only code, not metadata, it'll never save the resource name even if you set it. This is not the same for other types of resources, so it may be useful for those.

I don't know what you are expecting to get from the `resource_name` property of a script.

**Comment by damarindra** [comment]
Hmmm, If I remember correctly. I want to get the resource_name of the script to know if the node is Player or Enemy. So basically I have Script character_controller and Player and Enemy inherit from it. But I found a solution, I can set_meta() to know what is the node type exactly.

**Comment by vnen** [comment]
`get_script()` returns only the current script, which knows nothing about the node it's attached to (because the same script can be attached to multiple nodes).

**Comment by neikeq** [comment]
I can't check right now, but I think built-in scripts do use `resource_name`.

**Comment by mhilbrunner** [comment]
@vnen Should we just document this then and close this and https://github.com/godotengine/godot/pull/14774 as not a bug?

**Comment by vnen** [comment]
@mhilbrunner I'm not sure. #14774 uncomments a line that would fix this, but I'm not sure why it was commented out in the first place.

**Comment by akien-mga** [comment]
The plan for 3.2 is to outright get rid of `resource_name`.

**Comment by Anutrix** [comment]
So are we still getting rid of it for 3.2 or moved to 4.0?

**Comment by akien-mga** [comment]
4.0 I guess, there is little need to break compat in 3.2 for it as it's only cosmetic.

**Comment by somnivore** [comment]
I would actually prefer that `resource_name` be kept, since it serves a purpose in the inspector for local resources:

![image](https://user-images.githubusercontent.com/28724859/66464497-beb94780-ea44-11e9-9623-429a6d4d4113.png)


If it's removed, these resources would instead wind up looking like this:

![image](https://user-images.githubusercontent.com/28724859/66464558-dd1f4300-ea44-11e9-9d3f-0ff135fe4bc3.png)


As you can see, it's considerably more difficult to remember which resource is which, as I'm just memorizing some index on some array (which I'm probably not going to do and instead I'll wind up opening each resource one by one to get an idea of that resource's contents.)  I could just save each resource to disk to give them a name but the idea here is to not clutter the file system with these little resources (as there's going to be hundreds of them in this game, and all they really do is provide some name and description.)

I think it would be better to make it clear in `resource_name`'s tooltip that it's meant for editor-use only, rather than removing it all together.  Better yet is to re-enable `resource_name`'s functionality so I don't have to write `export var display_name := ""` in all my resources, and just exempt those resources which don't store `resource_name`, like `.gd` and `.cs` files, perhaps with a warning or error when attempting to get it.

**Comment by DarkMessiah** [comment]
I use `resource_name` in my custom editor and `resource_name` displays in inspector, when I call inspect_object(my_resource)
![image](https://user-images.githubusercontent.com/1756388/86525648-26a59d00-beb4-11ea-8e7d-3249d3dfed75.png)

**Comment by CarlGustavAlbertDwarfsteinYung** [comment]
I make use of resource_name to figure out what Material I am looking at. If you remove this I would have to parse resource_path to try and infer the material name and if no resource_path then I would need to get shader_param texture path and parse that and extract the name from there.  I would actually prefer the ability to set (optionally) set resource_name on any game resource.

**Comment by cgbeutler** [comment]
Looks like exporting a script with `export var script2 :Script` currently (3.2) results in this editor display:
![image](https://user-images.githubusercontent.com/9751923/102543877-2ddef280-4071-11eb-83a9-891370a5769e.png)

As noted in the bug, setting the resource_name of the script doesn't get saved. (It may show for a bit, till you restart the editor to clear the cache.)

If I'm not mistaken, the main issue is that the actual on-disk representation of the script is not a toml file, but just the script itself. Scripts have a custom importer that just uses the script itself. If I'm right, that would mean the resource_name would need to be saved somewhere in the script itself. Using `class_name` as the `resource_name` is a fairly obvious candidate. As a side note: Changing 'resource_name' _in_ the editor would still be problematic, as it would be way to hard to parse/edit reliably (due to comments and `\`, etc.)

TLDR;  Loading `class_name` into `resource_name` seems like a decent way to go.

**Comment by cgbeutler** [comment]
As a side note, I'm surprised that the `class_name` is not available anywhere on the [GDScript ](https://docs.godotengine.org/en/stable/classes/class_gdscript.html#class-gdscript)object. I would have expected it to be a field on the GDScript. I imagine if it isn't added as the resource_name, we may wanna make a feature request to add it. (Not sure it could be added to [Script](https://docs.godotengine.org/en/stable/classes/class_script.html#class-script) itself, as other scripting languages may not quite fit in that way.)
